File: simple_demo/client/hook_test.py
import sys
sys.path.append("../../../tm_transport/python_lib")

import os
import TMTransport
import asyncio
from enum import Enum
from typing import Callable
import cbor2 

import nacl.encoding
import nacl.signing
import nacl.secret
import nacl.utils
import nacl.hash
#from Crypto.Cipher import AES

async def asyncMain() :
    heartbeatLocator = 'rabbitmq://127.0.0.1::guest:guest:amq.topic[durable=true]'
    heartbeatTopic = 'simple_demo.secure_executables.calculator.heartbeat'

    serverPublicKey = b'\x69\x61\xB9\xCF\xBA\x37\xD0\xE2\x70\x32\x84\xF9\x41\x02\x17\x22\xFA\x89\x0F\xE4\xBA\xAC\xC8\x73\xB9\x00\x99\x24\x38\x42\xC2\x9A'
    decryptKey = nacl.hash.generichash('testkey'.encode('utf-8'), encoder=nacl.encoding.RawEncoder)[0:32]

    verifyKey = nacl.signing.VerifyKey(serverPublicKey)

    def verifyAndDecrypt(x : bytes) -> bytes:
        d = cbor2.loads(x)
        res = verifyKey.verify(d['data'], d['signature'])
        if not res:
            return None
        #l = int.from_bytes(res[0:8], byteorder='little')
        #cipher = AES.new(decryptKey, AES.MODE_CBC, iv=res[8:24])
        #plainText = cipher.decrypt(res[24:])
        #return plainText[0:l]
        nonce = res[0:nacl.secret.SecretBox.NONCE_SIZE]
        return nacl.secret.SecretBox(decryptKey).decrypt(res[nacl.secret.SecretBox.NONCE_SIZE:], nonce)

    tasks = []
    q = asyncio.Queue()
    tasks.extend(TMTransport.MultiTransportListener.input(heartbeatLocator, [q], heartbeatTopic, wireToUserHook=verifyAndDecrypt))
    async def printData():
        while True:
            topic, x = await q.get()
            print(f"{topic}: {repr(cbor2.loads(x))}")
    tasks.append(asyncio.create_task(printData()))
    await asyncio.gather(*tasks)

if __name__ == "__main__":
    asyncio.run(asyncMain())

File: simple_demo/client/hook_test.ts
import {MultiTransportListener} from '../../../tm_transport/node_lib/TMTransport'
import * as Stream from 'stream'
//import {eddsa as EDDSA} from "elliptic"
import * as cbor from 'cbor'
//import * as aes from 'aes-js'
import * as util from 'util'
import * as sodium from 'sodium-native'

const heartbeatLocator = 'rabbitmq://127.0.0.1::guest:guest:amq.topic[durable=true]'
const heartbeatTopic = 'simple_demo.secure_executables.calculator.heartbeat'

const serverPublicKey = Buffer.from([
    0x69,0x61,0xB9,0xCF,0xBA,0x37,0xD0,0xE2,0x70,0x32,0x84,0xF9,0x41,0x02,0x17,0x22,
    0xFA,0x89,0x0F,0xE4,0xBA,0xAC,0xC8,0x73,0xB9,0x00,0x99,0x24,0x38,0x42,0xC2,0x9A 
]);
/*let verifier = new EDDSA('ed25519');
const decryptKey = aes.utils.utf8.toBytes("testkey"+(' '.repeat(9)));*/
//const decryptKey = Buffer.from("testkey"+(' '.repeat(25)));
let decryptKey = Buffer.alloc(sodium.crypto_generichash_BYTES);
sodium.crypto_generichash(decryptKey, Buffer.from("testkey"));
decryptKey = decryptKey.slice(0, 32);

function verifyAndDecrypt(data : Buffer) : Buffer {
    let cborDecoded = cbor.decode(data);
    /*
    if (!verifier.verify(cborDecoded.data.toString('hex'), cborDecoded.signature.toString('hex'), serverPublicKey.toString('hex'))) {
        return null;
    }
    let len = Number((cborDecoded.data as Buffer).slice(0, 8).readBigUInt64LE());
    let decryptor = new aes.ModeOfOperation.cbc(decryptKey, cborDecoded.data.slice(8, 24));
    return Buffer.from(decryptor.decrypt(cborDecoded.data.slice(24))).slice(0, len);
    */
    if (!sodium.crypto_sign_verify_detached(
        cborDecoded.signature
        , cborDecoded.data
        , serverPublicKey
    )) {
        return null;
    }
    let ret = Buffer.alloc(cborDecoded.data.byteLength-sodium.crypto_secretbox_NONCEBYTES-sodium.crypto_secretbox_MACBYTES);
    if (sodium.crypto_secretbox_open_easy(
        ret
        , cborDecoded.data.slice(sodium.crypto_secretbox_NONCEBYTES)
        , cborDecoded.data.slice(0, sodium.crypto_secretbox_NONCEBYTES)
        , decryptKey
    )) {
        return ret;
    } else {
        return null;
    }
}

let heartbeatListener = MultiTransportListener.inputStream(
    heartbeatLocator
    , heartbeatTopic
    , verifyAndDecrypt
);

let outputStream = new Stream.Writable({
    write : function(chunk : [string, Buffer], _encoding, callback) {
        let content = cbor.decode(chunk[1]);
        let contentRep = util.inspect(content, { showHidden: false, depth: null, colors: true });
        console.log(`Topic ${chunk[0]}: ${contentRep}`);
        callback();
    }
    , objectMode : true
});

heartbeatListener.pipe(outputStream);